src/dbnode/storage/bootstrap/process.go

```diff
@@ -236,6 +242,32 @@ func (b bootstrapProcess) Run(
 		namespacesRunFirst,
 		namespacesRunSecond,
 	} {
+		for _, entry := range namespaces.Namespaces.Iter() {
+			ns := entry.Value()
+			// Check if snapshot-type ranges have advanced while bootstrapping previous ranges.
+			// If yes, return an error to force a retry
+			if persistConf := ns.DataRunOptions.RunOptions.PersistConfig(); persistConf.Enabled &&
+				persistConf.FileSetType == persist.FileSetSnapshotType {
+				var (
+					now                = b.nowFn()
+					nsOptions          = ns.Metadata.Options()
+					upToDateDataRanges = b.targetRangesForData(now, nsOptions.RetentionOptions())
+				)
+				// Only checking data ranges. Since index blocks can only be a multiple of
+				// data block size, the ranges for index could advance only if data ranges
+				// have advanced, too (while opposite is not necessarily true)
+				if !upToDateDataRanges.secondRangeWithPersistFalse.Range.Equal(ns.DataTargetRange.Range) {
+					upToDateIndexRanges := b.targetRangesForIndex(now, nsOptions.RetentionOptions(),
+						nsOptions.IndexOptions())
+					fields := b.logFields(ns.Metadata, ns.Shards,
+						upToDateDataRanges.secondRangeWithPersistFalse.Range,
+						upToDateIndexRanges.secondRangeWithPersistFalse.Range)
+					b.log.Error("time ranges of snapshot-type blocks advanced", fields...)
+					return NamespaceResults{}, ErrFileSetSnapshotTypeRangeAdvanced
```

Comment:
Are the INITIALIZING shards still INITIALIZING on retry? Maybe we can add/modify an integration test for this - would like to validate that it shards are still INITIALIZING on retry and it properly bootstraps on the retry.

Reply:
Added a couple of tests for bootstrap retries https://github.com/m3db/m3/pull/3223/commits/ec1999639d4e398d2be51d1416f341f61cd47e5d

Couldn't capture a shard state change from `INITIALIZING` to `AVAILABLE`, for some unknown reason, shards in these tests have `AVAILABLE` state even before boostrapping finishes the first pass. I'm checking the bootstrapped state of DB, this seems to behave as expected.

Reply:
Synced offline - shards should still be INITIALIZING on retry (peers bootstrapping). 

Just sharing links from the chat here on how integration test setup placement/topology:
https://github.com/m3db/m3/blob/master/src/dbnode/integration/serve.go#L52
https://github.com/m3db/m3/blob/master/src/dbnode/integration/setup.go#L305
https://github.com/m3db/m3/blob/master/src/dbnode/integration/setup.go#L322

It seems like it'll be a non significant amount of work to make the integration tests more closely w/ reality and add a more intricate test here I think the ones you added are fine.